# Problemas de prueba de hipótesis

## Ejercicio 1

Usted es un nutricionista que investiga dos tipos diferentes de dietas para ver si existe una diferencia significativa en la pérdida de peso después de un mes. Eliges dos grupos aleatorios de personas; un grupo sigue la primera dieta y el otro sigue la segunda. Al final del mes, se registra la pérdida de peso (en kg) de cada persona:

| Diet 1 | Diet 2 |
|:-------|:-------|
| 2.0 | 3.0 |
| 2.5 | 3.2 |
| 3.0 | 3.1 |
| 2.8 | 2.9 |
| 2.3 | 2.8 |
| 2.7 | 3.0 |
| 2.5 | 3.2 |

Con estos datos se busca responder a la siguiente pregunta: ¿Existe una diferencia significativa en la pérdida de peso promedio entre las personas que siguieron la primera dieta y las que siguieron la segunda dieta?

Para sacar conclusiones, siga los siguientes puntos:

-Enunciar la hipótesis: hipótesis nula y alternativa.
-Realizar la prueba para comprobar la hipótesis. Puede utilizar la prueba t de Student.
-Analizar las conclusiones.

In [10]:
import numpy as np
from scipy import stats

# ============================================================
# EJERCICIO 1 – Comparación de dos grupos (t de Student)
# ============================================================

alpha = 0.05

# -----------------------------
# 1. Datos 
# -----------------------------

grupo_a = np.array([2.0, 2.5, 3.0, 2.8, 2.3, 2.7, 2.5])
grupo_b = np.array([3.0, 3.2, 3.1, 2.9, 2.8, 3.0, 3.2])


print("Tamaño muestra grupo A:", len(grupo_a))
print("Tamaño muestra grupo B:", len(grupo_b))
print("Media grupo A:", np.mean(grupo_a))
print("Media grupo B:", np.mean(grupo_b))
print("Desviación típica grupo A:", np.std(grupo_a, ddof=1))
print("Desviación típica grupo B:", np.std(grupo_b, ddof=1))

# -----------------------------
# 2. Hipótesis
# -----------------------------
# H0: μ_A = μ_B
# H1: μ_A ≠ μ_B

# -----------------------------
# 3. Shapiro-Wilk
# -----------------------------
stat_a, p_a = stats.shapiro(grupo_a)
stat_b, p_b = stats.shapiro(grupo_b)

print("\n--- Normalidad Shapiro-Wilk ---")
print(f"Grupo A: estadístico={stat_a:.4f}, p={p_a:.4f}")
print(f"Grupo B: estadístico={stat_b:.4f}, p={p_b:.4f}")

normal_a = p_a > alpha
normal_b = p_b > alpha

if normal_a and normal_b:
    print("\nAmbos grupos son aproximadamente normales.")
else:
    print("\nAl menos un grupo NO es normal. El t-test no es perfecto aquí.")

# -----------------------------
# 4. Comprobación de homogeneidad de varianzas (Levene)
# -----------------------------
stat_levene, p_levene = stats.levene(grupo_a, grupo_b)
print("\n--- Test de Levene (igualdad de varianzas) ---")
print(f"Estadístico = {stat_levene:.4f}, p-valor = {p_levene:.4f}")

varianzas_iguales = p_levene > alpha

if varianzas_iguales:
    print("Podemos asumir varianzas iguales (se usa t-test clásico).")
else:
    print("No podemos asumir varianzas iguales (se usará t-test de Welch).")

# -----------------------------
# 5. t-test (solo si hay normalidad)
# -----------------------------
if normal_a and normal_b:
    if varianzas_iguales:
        # t de Student clásico (varianzas iguales)
        t_stat, p_val = stats.ttest_ind(grupo_a, grupo_b, equal_var=True)
        print("\n--- t de Student con varianzas iguales ---")
    else:
        # t de Welch (varianzas desiguales)
        t_stat, p_val = stats.ttest_ind(grupo_a, grupo_b, equal_var=False)
        print("\n--- t de Welch (varianzas desiguales) ---")

    print(f"t = {t_stat:.4f}, p = {p_val:.4f}")

    if p_val < alpha:
        print(f"\nSe rechaza H0 (p={p_val:.4f} < {alpha}). Hay diferencias significativas.")
    else:
        print(f"\nNo se rechaza H0 (p={p_val:.4f} >= {alpha}). No hay diferencias significativas.")
else:
    print("\n**No se ejecuta t-test porque no hay normalidad en ambos grupos.**")


Tamaño muestra grupo A: 7
Tamaño muestra grupo B: 7
Media grupo A: 2.542857142857143
Media grupo B: 3.0285714285714285
Desviación típica grupo A: 0.3309438162646487
Desviación típica grupo B: 0.14960264830861927

--- Normalidad Shapiro-Wilk ---
Grupo A: estadístico=0.9808, p=0.9635
Grupo B: estadístico=0.9346, p=0.5905

Ambos grupos son aproximadamente normales.

--- Test de Levene (igualdad de varianzas) ---
Estadístico = 2.2710, p-valor = 0.1577
Podemos asumir varianzas iguales (se usa t-test clásico).

--- t de Student con varianzas iguales ---
t = -3.5383, p = 0.0041

Se rechaza H0 (p=0.0041 < 0.05). Hay diferencias significativas.


La prueba t de Student para muestras independientes es apropiada cuando se desea
comparar la media de dos grupos independientes bajo el supuesto de que ambos
proceden de poblaciones con distribución normal y varianzas similares.


Conclusión Ejercicio 1

Primero se comprobaron los supuestos necesarios para aplicar una prueba t de Student para muestras independientes:

Normalidad:
Los dos grupos presentan p-valores > 0.05 en la prueba de Shapiro–Wilk, por lo que no se rechaza la hipótesis de normalidad en ninguno de ellos.

Igualdad de varianzas:
El test de Levene arrojó un p-valor = 0.1577 (> 0.05), lo que indica que no se rechaza la hipótesis de homogeneidad de varianzas.
Por tanto, es adecuado utilizar la t de Student clásica (con varianzas iguales).

Dado que se cumplen los supuestos, se aplicó una prueba t de Student para dos muestras independientes con varianzas iguales.

La prueba arrojó:

Estadístico t = –3.5383

p-valor = 0.0041

Como p-valor < 0.05, se rechaza la hipótesis nula (H₀).
Por tanto, existen diferencias estadísticamente significativas en la pérdida de peso promedio entre las dos dietas.

A continuación, se compararon las medias de cada grupo:

Dieta 1: media = 2.54 kg

Dieta 2: media = 3.03 kg

El grupo con la mayor pérdida de peso promedio es el GRUPO 2 (Dieta 2).

## Ejercicio 2

Un agricultor decide probar tres tipos diferentes de fertilizantes para determinar si alguno es superior en términos de producción de maíz. El agricultor planta maíz en 15 parcelas idénticas y utiliza los tres fertilizantes (5 parcelas para cada tipo). Al final de la temporada mide el rendimiento de maíz (en kg) de cada parcela, con el siguiente resultado:

| Fertilizante 1 | Fertilizante 2 | Fertilizante 3 |
|:-------------|:-------------|:-------------|
| 20 | 22 | 24 |
| 21 | 21 | 23 |
| 20 | 23 | 22 |
| 19 | 22 | 23 |
| 20 | 21 | 24 |

Con estos datos busca responder la siguiente pregunta: ¿Existe una diferencia significativa en el rendimiento promedio del maíz entre los tres tipos de fertilizantes?

Para ayudarte, sigue los siguientes puntos:

-Enunciar la hipótesis: hipótesis nula y alternativa.
-Realizar la prueba ANOVA.
-Analizar las conclusiones.
-Si un fertilizante es mejor que otro, ¿cómo podemos saberlo?

1) Datos y descriptivos

In [11]:
import numpy as np
from scipy import stats
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# ============================================================
# EJERCICIO 2 – Comparación de más de dos grupos
# ============================================================

alpha = 0.05

# ------------------------------------------------------------
# 1. Datos (CORREGIDOS SEGÚN EL ENUNCIADO)
# ------------------------------------------------------------

fert_1 = np.array([20, 21, 20, 19, 20])
fert_2 = np.array([22, 21, 23, 22, 21])
fert_3 = np.array([24, 23, 22, 23, 24])

grupos = [fert_1, fert_2, fert_3]

# ------------------------------------------------------------
# 2. Pruebas de normalidad (Shapiro-Wilk)
# ------------------------------------------------------------
print("Prueba de normalidad (Shapiro-Wilk):")
for i, g in enumerate(grupos, start=1):
    stat, p = stats.shapiro(g)
    print(f"  Grupo {i}: p-value = {p:.4f}")

# ------------------------------------------------------------
# 3. Homogeneidad de varianzas (Levene)
# ------------------------------------------------------------
stat, p_levene = stats.levene(fert_1, fert_2, fert_3)
print(f"\nPrueba de Levene: p-value = {p_levene:.4f}")

# ------------------------------------------------------------
# 4. ANOVA de una vía
# ------------------------------------------------------------
f_stat, p_anova = stats.f_oneway(fert_1, fert_2, fert_3)
print(f"\nANOVA: p-value = {p_anova:.4f}")

# ------------------------------------------------------------
# 5. Tukey HSD si procede
# ------------------------------------------------------------
if p_anova < alpha:
    print("\n--- Tukey HSD ---")
    datos = np.concatenate(grupos)
    etiquetas = (["F1"] * len(fert_1) +
                 ["F2"] * len(fert_2) +
                 ["F3"] * len(fert_3))
    
    tukey = pairwise_tukeyhsd(endog=datos, groups=etiquetas, alpha=alpha)
    print(tukey)
else:
    print("\nNo se realiza Tukey HSD (no hay diferencias globales significativas).")

# ------------------------------------------------------------
# 6. ¿Qué grupo es mejor? (mayor media)
# ------------------------------------------------------------
medias = [np.mean(g) for g in grupos]
print("\nMedias de los grupos:")
for i, m in enumerate(medias, start=1):
    print(f"  Grupo {i}: media = {m:.2f}")

indice_mejor = int(np.argmax(medias))
print(f"\nEl fertilizante con mayor producción media es el **GRUPO {indice_mejor + 1}**.")


Prueba de normalidad (Shapiro-Wilk):
  Grupo 1: p-value = 0.3254
  Grupo 2: p-value = 0.3140
  Grupo 3: p-value = 0.3140

Prueba de Levene: p-value = 0.8040

ANOVA: p-value = 0.0001

--- Tukey HSD ---
Multiple Comparison of Means - Tukey HSD, FWER=0.05
group1 group2 meandiff p-adj  lower  upper  reject
--------------------------------------------------
    F1     F2      1.8 0.0099 0.4572 3.1428   True
    F1     F3      3.2 0.0001 1.8572 4.5428   True
    F2     F3      1.4 0.0409 0.0572 2.7428   True
--------------------------------------------------

Medias de los grupos:
  Grupo 1: media = 20.00
  Grupo 2: media = 21.80
  Grupo 3: media = 23.20

El fertilizante con mayor producción media es el **GRUPO 3**.


El ANOVA de una vía es el test adecuado cuando se desea comparar las medias
de **tres o más grupos independientes** y se cumplen los supuestos de:

1. Normalidad en cada grupo.
2. Homogeneidad de varianzas entre los grupos.
3. Independencia de las observaciones.

Dado que estos supuestos se verifican en los datos analizados, el ANOVA es la
prueba estadística apropiada.


Conclusión Ejercicio 2

Primero se comprobaron los supuestos necesarios para aplicar un ANOVA de una vía:

Normalidad:
Los tres grupos presentan p-valores > 0.05 en la prueba de Shapiro–Wilk, por lo que no se rechaza la hipótesis de normalidad en ninguno de ellos.

Igualdad de varianzas:
El test de Levene arrojó un p-valor = 0.8040 (> 0.05), lo que indica que no se rechaza la hipótesis de homogeneidad de varianzas.
Por tanto, las tres poblaciones pueden considerarse con varianzas iguales.

Dado que se cumplen los supuestos, se aplicó un ANOVA de una vía (paramétrico) para comparar las medias de los tres grupos.

El ANOVA arrojó:

Estadístico F (calculado por la librería)

p-valor = 0.0001

Como p-valor < 0.05, se rechaza la hipótesis nula (H₀).
Por tanto, existen diferencias estadísticamente significativas entre al menos dos de los tres grupos.

A continuación, se compararon las medias de cada grupo:

Grupo 1: media = 20.00

Grupo 2: media = 21.80

Grupo 3: media = 23.20

El grupo con el mejor resultado promedio es el GRUPO 3, al ser el que presenta la media más alta.

Desde un punto de vista práctico, los resultados indican que el fertilizante del
Grupo 2 permite obtener una producción significativamente mayor que los otros
dos fertilizantes analizados. Esto lo convierte en la opción más efectiva bajo
las condiciones del experimento.
